### ISSUE TITLE
Replace UUID string with a RFC compliant standard form UUID [13]

### ISSUE THREAD

**Issue by MartinWahnschaffe** [SORMAS-Foundation/SORMAS-Project#7086]
### Feature Description
SORMAS generates v4 UUIDs, and stores them as a Base32-encoded string (e.g. TFRGBU-UVB25S-VSUAFH-QLYFKKOA). The first 6 characters are used as a short ID in tables and other overviews. The reasoning behind this was to shorten the uuid a bit and to have a break after 6 characters that is easy to identify and allows us to use the first 6 characters as a short id.

### Problem Description
Not using the standard form (e.g. a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11) makes it more difficult to interoperate with third party systems. In addition we can't use the UUID database and Java type, which would be much more convenient (see https://www.postgresql.org/docs/current/datatype-uuid.html).
Also the current base32 implementation is not RFC compliant. See #6706 

### Proposed Change

Depends on https://github.com/hzi-braunschweig/SORMAS-Project/issues/10249. Having  #9708 done before that would greatly reduce refactoring overhead.

- [ ] In the database rename all uuid columns to legacyUuid. Remove the not null constraint
- [ ] In the database create a new unique uuid column for each table, using the Postgres UUID type. 
- [ ] Enable the UUID-OSSP module for postgres, by adding this to the server-setup.sh: `CREATE EXTENSION IF NOT EXISTS "uuid-ossp";`
- [ ] Add similar section to the server-update.sh (clone the basics to run SQL from server-setup.sh)
- [ ] Fill the new uuid column by creating a v5 uuid using the legacyUuid as input. Use uuid_generate_v5 from https://www.postgresql.org/docs/current/uuid-ossp.html. In order to do so, the module may have to be activated in the server setup/update scripts. 
**Namespace**: A v5 UUID also needs to have a namespace uuid. For that use a NIL UUID `00000000-0000-0000-0000-000000000000` (Java: `new UUID( 0 , 0 );`, Postgres: `uuid_nil()`)
**Important**: Infrastructure entities with the centrallyManaged flag set, need to be excluded from this - they already have a correct UUID provided by central.
- [ ] Use the same logic to create the uuid for "constant" uuids as used for "other health facility" and similar. I suggest to add a `DataHelper.createUuidFromLegacy` method that uses a library like https://github.com/f4b6a3/uuid-creator/ to generate the v5 UUID - unfortunately Java does not provide a method for this.
- [ ] New UUIDs should be created using v4 `UUID.randomUUID()`.
- [ ] Change the data type of the uuids in all Java classes to java.util.UUID: abstract entities, dtos, criterias, getByUuid facade methods, a few more methods (e.g. AuditLoggerFacade.logExternalMessagesHtmlSuccess)
- [ ] Introduce a new field "legacyUuid" for all entities (`AbstractDomainObject`) and entityDtos
- [ ] When new entites are created no longer fill the legacyUUID
- [ ] Apply the same changes to the android app and it's backend
- [ ] Make sure the open API swagger definition uses `type: string` in combination with `format: uuid` for the new uuid
- [ ] When an entity is opened based on an url and no result is found also check the legacy URL
- [ ] Extend buildCriteriaFilter methods to also include the legacy UUID for "like" search of Person, Case, Contact, Sample, Task, TravelEntry, User, Immunization, ExternalMessage, Event, EventGroup
- [ ] Display the legacy UUID in the edit form of all core entities
- [ ] Add a configuration to disabled displaying the legacy UUID
- [ ] #10364 introduced `AbstractUuidDto`. Check if other big base classes like e.g., `EntityDto`, `ReferenceDto` etc. can extend it to make refactoring easier.
- [ ] `AbstractUuidDto.uuid` should probably be required/not-null. 

#### External Systems

- [x] Data synchronized via central already has correct UUIDs
- [x] Create issue: The survnet converter needs to be updated to work with the new UUID data type instead of the base 32 uuid string and rolled-out together with the new SORMAS version. There is no need to migrate data.
-> https://github.com/hzi-braunschweig/sormas-survnet-interface/issues/361
- [x] Create issue: The DEMIS Adapter needs to be checked and probably updated to work with the new UUID data type instead of the base 32 uuid string and rolled-out together with the new SORMAS version. There is no need to migrate data. Looks like only the new sormas-api version needs to be used and the LabMessageConverterUnitTest needs to be adjusted: https://git.vitasystems.dev/search?search=uuid&nav_source=navbar&project_id=780&group_id=165&search_code=true&repository_ref=development
-> https://github.com/hzi-braunschweig/sormas-demis-adapter/issues/84
- [ ] Provide a short migration info for external systems that will be part of the release notes. Should contain a Java and a Postgres code snippet on how to generate a v5 UUID based on the legacy UUID.
- [ ] The changes should be announced one iteration before they are released

### Additional information

* A url-safe short UUID (e.g. based on flickr58 alphabet - see https://www.npmjs.com/package/short-uuid) would be handy at some point, but is probably not urgently needed and thus not part of this issue.
* Sequential uuids would be beneficial for database performance but won't work with the way data creation is distributed in SORMAS (mobile apps, external systems).
* Optional: When a dto does have a legacy UUID, but no UUID, retrieve the UUID based on the legacy UUID?

**Comment by MartinWahnschaffe** [comment]
@leventegal-she Can you check whether is any functionality related to S2S that needs to be taken into account for this issue?
@JonasCir Please also have another look at the refined issue.

**Comment by MartinWahnschaffe** [comment]
@JaquM-HZI This issue will change the UUID of cases and all other entities. The old UUID will be kept as legacy UUID.

![grafik](https://user-images.githubusercontent.com/23217632/183579975-8fafda49-a3d5-4939-9f21-63e66122378c.png)

Is there a need to display the legacy UUID to users for a transition period (e.g. three releases or via configuration)?

**Comment by JaquM-HZI** [comment]
@MartinWahnschaffe yes it is necessary to be able to see them or have them in an export, cause sometimes it is used for official documents etc. 
Can't say for how long, but would say at least 6 months, so please make it available for this time minimum.

**Comment by JonasCir** [comment]
@MartinWahnschaffe thank you very much for pushing this. Issue looks good to me, I added my thoughts to the refine section.

**Comment by MartinWahnschaffe** [comment]
@JonasCir Can you provide details on what needs to be done for the central etcd process when the uuid is changed? I guess all infrastructure data on the etcd has to be updated at the same time the new version is rolled out. This means the version has to be rolled out to all servers using the etcd at the same time, right?

**Comment by JonasCir** [comment]
Goog catch @MartinWahnschaffe! I was thinking ahead and implemented central in such a way that it only distributes valid UUIDv5. See [here](https://github.com/hzi-braunschweig/SORMAS-central-data/blob/main/main.py#L24) for an example of how existing data can be fed into `uuidv5` to generate a stable identifier.

**Comment by MartinWahnschaffe** [comment]
@JonasCir Hm, unfortunately this makes the whole thing more complicated. It means that we have to check whether a legacy UUID string already is a valid UUID and need to keep it in this case. The same thing has to be done by external systems that rely on the SORMAS UUIDs.

**Comment by JonasCir** [comment]
@MartinWahnschaffe and  I discussed the last point: infrastructure data has an `centrallyManaged` flag. If it is set, the UUID is valid and does not need to be recomputed.
